From 75b1a71dc69abb5e1c1982f7a64ad9027210a85d Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 25 Jul 2018 15:42:51 +0800
Subject: [PATCH 4275/5242] MLK-19017-7 drm/imx: sec-dsim_imx: add rpm status
 check for suspend/resume

commit  f9506f8686800b764d6e6d9555a258835a61b5a8 from
https://source.codeaurora.org/external/imx/linux-imx.git

Add runtime PM status check during runtime suspend and resume
to avoid unnecessary jobs if it is already in that state which
can avoid possible kernel warnings of clock disable/unprepare
mismatch during system suspend if it is alreay in runtime
suspended state.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 6f95c6fdc0de2fd4fe1d835c164f5e3cfb23e17d)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/imx/sec_mipi_dsim-imx.c |   13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c b/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
index e57451d..bc985dd 100644
--- a/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
+++ b/drivers/gpu/drm/imx/sec_mipi_dsim-imx.c
@@ -49,6 +49,8 @@ struct imx_sec_dsim_device {
 	struct device *dev;
 	struct drm_encoder encoder;
 	struct regmap *gpr;
+
+	bool rpm_suspended;
 };
 
 #define enc_to_dsim(enc) container_of(enc, struct imx_sec_dsim_device, encoder)
@@ -229,6 +231,7 @@ static int imx_sec_dsim_bind(struct device *dev, struct device *master,
 	}
 
 	pm_runtime_enable(dev);
+	dsim_dev->rpm_suspended = true;
 
 	dev_dbg(dev, "%s: dsim bind end\n", __func__);
 
@@ -288,15 +291,23 @@ static int imx_sec_dsim_resume(struct device *dev)
 #ifdef CONFIG_PM
 static int imx_sec_dsim_runtime_suspend(struct device *dev)
 {
+	if (dsim_dev->rpm_suspended == true)
+		return 0;
+
 	sec_mipi_dsim_suspend(dev);
 
 	release_bus_freq(BUS_FREQ_HIGH);
 
+	dsim_dev->rpm_suspended = true;
+
 	return 0;
 }
 
 static int imx_sec_dsim_runtime_resume(struct device *dev)
 {
+	if (dsim_dev->rpm_suspended == false)
+		return 0;
+
 	request_bus_freq(BUS_FREQ_HIGH);
 
 	/* Pull dsim out of reset */
@@ -306,6 +317,8 @@ static int imx_sec_dsim_runtime_resume(struct device *dev)
 
 	sec_mipi_dsim_resume(dev);
 
+	dsim_dev->rpm_suspended = false;
+
 	return 0;
 }
 #endif
-- 
1.7.9.5

